Add engineering career development framework (#3954)
sourcegraph/about: handbook/engineering/career-development/talent-review-process.md

#### Considerations for promotion

Teammates are considered for promotion to a new level when their manager can make the case that they have had at least one quarter of high performance at their current level, and one quarter performing at the next level. We want to avoid situations where someone is promoted to a level in which they struggle to meet expectations.

The engineering levels describe performance expectations in 3 categories. It will be extremely rare for an engineer to be exactly at the same level in all of these categories. Since the levels describe the minimum expectations for each level, an engineer must meet the expectations for all categories before they can be considered for promotion.

This does imply a distinction between _high performance at current level_ vs. _ready for promotion_. A level N engineer who is consistently exceeding level N expectations in some areas, but whom we do not feel would be able to meet level N+1 expectations, should not be promoted (despite their high performance in some areas).

> 
> ## FAQ
> 
> ### Why are we even doing this?
> 
> See these [problems we are feeling today](https://docs.google.com/document/d/1KV0eq3YKFHLTiak_zmbzmjhvZ5o1odjITazY2kW6KvA/edit#heading=h.3hxcl9aecpbb).
> 
> ### Why is talent review done quarterly?
> 
> More frequent than quarterly is not enough time for significant growth to occur for most ICs, and is thus not a good use of EM effort: too much cost for not enough benefit. Less frequent than quarterly raises the stakes too high: “I just started this new role right before Talent Review. If it’s not enough time for me to demonstrate my increased impact, I’ll have to wait another 6 months before I’m considered again!”
> 
> ### So I just have to do what’s listed in the level definition to be promoted to that level?
> 
> No, for several reasons. First, the levels are cumulative, so the expectations of previous levels still hold. Second, we expect everyone to exhibit our values at all levels (but did not list this in the levels). But most importantly, the level definitions are _examples_ of the kinds of behaviors and impact we expect of someone operating at that level. The exact examples provided are neither necessary nor sufficient for promotion.
> 
> ### How did you come up with all of this?
> 
> Glad you asked! We started with [research](https://docs.google.com/document/d/1Vqpks_iZLalGwbcTFHQcQs3xnjZtGxX_azH063aulOg/edit#).

Reviewer: ```suggestion
```

This is done for each of our engineering levels.

## After the meeting

Decisions regarding promotions are prioritized by the relevant decision-makers (EM, Director, VP), and made quickly. Promotions are communicated back to the ICs in a timely manner. Regardless of promotions, any qualitative feedback for an IC that came out of talent review is shared with them in their next one-on-one (if not sooner).

It’s important that this process moves quickly! If we leave people wondering for a month or more if they will be getting a promotion, we’re creating needless anxiety and uncertainty.

Finally, we should have a [retrospective](../../retrospectives/index.md) after each of these, so we can continue iterating and improving upon this process. We should also consider changes/clarifications to the [level descriptions](framework.md#levels) at this time, in case any of the existing definitions were cause for confusion.

> ## Levels

Reviewer 2: much of the content in this section feels more relevant to the page that documents the levels themselves rather than the talent review process. There is also some duplication that I think would be nice to remove.

Reviewer: I will remove some of the duplication tomorrow, but I think we might merge what we have first, before Felix goes on break.

Author: Thanks for the feedback! @chrispine will do some deduplication tomorrow

#### Research

We consulted a variety of sources to determine our levels, and [summarized our findings here](https://docs.google.com/document/d/1Vqpks_iZLalGwbcTFHQcQs3xnjZtGxX_azH063aulOg/edit#).

#### Evaluation categories

Each level definition is composed of three categories, each with a summary statement and several example behaviors.

The categories we use are:

- Proficiency
- Execution
- Teamwork

#### Level descriptions

The levels are [defined in our career development framework](framework.md). In most cases, a level builds on the criteria from the preceding level: someone at level 2 must also meet the criteria for level 1.

In addition to what is listed there, we expect engineers at all levels to exhibit [our values](../../company/values.md).

These are the expectations for ICs _after they have completed their onboarding_. Some of these expectations (such as around communication) would start on day 1, but others (such as expertise in the codebase) would not be expected until they are fully ramped up.

Rather than precede each bullet point in the level descriptions with “Consistently”, we leave it as implicit.

> In line with our ["Continuously grow" company value](../../company/values.md), we expect every engineer to eventually reach at least level 3, spending no more than 2 years at any previous level. It is the responsibility of the EM to track this, and to ensure that ICs are given the support and opportunities needed for career growth.

Reviewer 2: duplicated

Reviewer 2: ```suggestion
```

This is done for each of our engineering levels.

## After the meeting

Decisions regarding promotions are prioritized by the relevant decision-makers (EM, Director, VP), and made quickly. Promotions are communicated back to the ICs in a timely manner. Regardless of promotions, any qualitative feedback for an IC that came out of talent review is shared with them in their next one-on-one (if not sooner).

It’s important that this process moves quickly! If we leave people wondering for a month or more if they will be getting a promotion, we’re creating needless anxiety and uncertainty.

Finally, we should have a [retrospective](../../retrospectives/index.md) after each of these, so we can continue iterating and improving upon this process. We should also consider changes/clarifications to the [level descriptions](levels.md#levels) at this time, in case any of the existing definitions were cause for confusion.

<!-- Note from Felix: Everything below here feels in large part duplicative now from levels.md ? -->

## Levels

#### Structure of the levels

> We have 7 levels for software engineers at Sourcegraph, derived from Option Impact, a startup compensation tool that Sourcegraph and many other companies use to determine their compensation bands. For each of these, we have a corresponding level definition. (Note: Initially, we are only defining levels 1-5.)
> 
> The structure of a level definition is in three parts, for each of our evaluation categories (listed below). For each category, we include a summary statement and several _examples_ of the kinds of behaviors we expect of someone at that level. (Note that these are neither necessary nor sufficient for promotion: these are not checkboxes. Rather, they serve to illustrate the magnitude of impact that we are looking for in each of the categories.)
> 
> The level definitions correspond to the start of the level, so that if an IC has a level N impact in all categories, they should then be promoted to level N (and not before). Or to put it another way, the level definitions describe the minimum expectations.

Reviewer: ```suggestion
We have 7 levels for software engineers at Sourcegraph, derived from Option Impact, a startup compensation tool that Sourcegraph and many other companies use to determine their compensation bands. For each of these, we have a corresponding level description.

The structure of a level description is in three parts, for each of our evaluation categories (listed below). For each category, we include a summary statement and several _examples_ of the kinds of behaviors we expect of someone at that level. (Note that these are neither necessary nor sufficient for promotion: these are not checkboxes. Rather, they serve to illustrate the magnitude of impact that we are looking for in each of the categories.)

The level descriptions correspond to the start of the level, so that if an IC has a level N impact in all categories, they should then be promoted to level N (and not before). Or to put it another way, the level descriptions state the minimum expectations.
```

#### Research

We consulted a variety of sources to determine our levels, and [summarized our findings here](https://docs.google.com/document/d/1Vqpks_iZLalGwbcTFHQcQs3xnjZtGxX_azH063aulOg/edit#).

#### Evaluation categories

Each level definition is composed of three categories, each with a summary statement and several example behaviors.

The categories we use are:

- Proficiency
- Execution
- Teamwork

> #### Level definitions

Reviewer: ```suggestion
#### Level descriptions
```
